### DISCUSSION TITLE
Membership of hierarchical groups

### DISCUSSION THREAD

**Discussion by KirschbaumP** [keycloak/keycloak#12726] [Feedback]
Corresponding to https://github.com/keycloak/keycloak/issues/9482, I would like to make a suggestion for improvement on hierarchical groups and start a discussion about it.

### Actual state
You have a _supergroup1_ and a _supergroup1/subgroup1_.

If a user is member of _supergroup1/subgroup1_ and you try to add a user to _supergroup1_ you get an 200 (OK) response, but keycloaks doesn't add the user to _supergroup1_ because the user is already an indirect member of it.

If you first add the user to _supergroup1_ and then to _supergroup1/subgroup1_, the user is member of both groups.

### Enhancement
In any case the api should not respond with 200 if the user was not added to the group if it is an (indirect) member of it.

#### Suggestion 1
Create a configuration parameter for each realm to control whether the user can be added to a group of which he is already an indirect member.

#### Suggestion 2
Add an parameter _ignoreIndirectMembership_ to the "_PUT /{realm}/users/{id}/groups/{groupId}_" Endpoint. The parameter should be an optional parameter, with the default value _false_, so that no existing implementations are damaged.

**Comment by u3r** [comment]
I'd like to add 
**Suggestion 3:**
Fix the test when adding a user to a group to only check if the user is direct member to that group!
I believe there is a method there that mixes two different purposes: one of checking whether a user has access (by transitive membership) and one preventing duplicate assignment.

Even reading @hmlnarik 's comment https://github.com/keycloak/keycloak/issues/9482#issuecomment-1162891811 I still guess this was initially an accident.

**If** it is NOW considered an "intended behaviour" (even though inconsistant based on order) I need to state:
I think this design decision is questionable from a corporate usage point of view:
At least at my company it is very common for people to get access to certain resources from "different sources" 

**Example:**
*Given:*
I have access to some repository because I am a developer in that project, and I also have access to that same repository because I am a specialist in a certain topic and also because I have some kind of controlling function
*When:* I change teams 
Thereby loosing the first "kind" of access
*Then:*
 I should STILL have access to that repository via the other two functions I hold.
 This would be **prevented** (depending on the order in which I get this hierarchical Access) in the current code base.

Also I don't see where this simplifies anything. Neither in Keycloaks code nor in any of the handling code (based on Suggestion 1 and Suggestion 2) nor any mental model.

See also (https://issues.redhat.com/browse/KEYCLOAK-10603) and https://github.com/keycloak/keycloak/pull/7970 for a very similar issue with users and roles which has been fixed.

**Comment by u3r** [comment]
Having checked the code (because I assumed a reason similar to that of #7970 (as solution to the original issue [KEYCLOAK-10603](https://issues.redhat.com/browse/KEYCLOAK-10603))):
The current behaviour has indeed been in the first commit supporting groups.

Due to the issues arising from different ordering of group assignment (first to sub-group then super vs. first to super-group and then to sub-group) not being idempotent I would still see this as a bug and propose to treat it the same as role/user assignment:

Split checks into access-check --> transitively checks
and membership-assignment --> only checks the current group.

**Comment by stianst** [comment]
In the current implementation if a user is member of a child group first, then added to a parent group, the membership for the child group is not removed. As it's no longer possible to add the user to another child group after it is added to the parent this leads to inconsistencies. In addition the lack of error when trying to add a user to a child group is not good.

I see two ways this can be handled:

1. Not allow a user to be member of a child group. This would require removing direct child group memberships as a user is added to a parent group, and should also return an error (instead of just silent ignoring) when trying to add a user to a child group where it is already a member of a parent.
2. Just allow adding direct membership to a child group even if the user is member of a parent group.

I don't honestly see any issues with 2, and it is significantly easier to update the behaviour to this than it would be to do 1. This will also allow adding direct child group membership, which still takes affect if membership to a parent is removed, which I can see would be a fairly common thing to want to do.

**Reply by hmlnarik** [reply]
I agree with 2. The problem with Option 1 I see is exactly the same - there is no easy way to "undo" adding user to a parent group when they were previously member of a child group(s).

**Reply by leischt** [reply]
I also think option 2 would create a understandable and consistent behavior. I created a PR, which implements option 2 as part of the issue this discussion was start for. Could you please have a look into this and let me know if this is the right direction?
Issue: [9482](https://github.com/keycloak/keycloak/issues/9482)
PR: [10785](https://github.com/keycloak/keycloak/pull/10785)

**Reply by apanagio** [reply]
I don't know if this is an old post. I am using 19.02 and the behavior I see is that a user can be added to a child group even if the user is a member of a parent group.
What cannot be done is add the user to the parent group when the user is a member of a child group

**Reply by apanagio** [reply]
Actually using the UI you cannot add a user to a child group if you go through the user.
However if you go to the child group you can add the user, even if the user is a member of the parent group.
I think this is an inconsistent behavior.

**Reply by sschu** [reply]
The fix will only be included in Keycloak 20. @apanagio
